app/services/smart_proxy_reference.rb

```diff
@@ -0,0 +1,49 @@
+module SmartProxyReference
+  class << self
+    attr_accessor :references
+
+    def smart_proxy_reference(hash)
+      raise Foreman::Exception.new(N_("Proxy reference should be { :relation => [:column_name, ...] }")) unless correct_reference_format?(hash)
+      hash.deep_clone
+    end
+
+    def add_smart_proxy_reference(hash)
+      hash_clone = smart_proxy_reference hash
+      if @references
+        join_references @references, hash_clone
+      else
+        @references = hash_clone
+      end
+    end
+
+    def smart_proxy_references
+      result = local_entries.deep_clone
+      join_references result, entries_from_plugins
+    end
+
+    def local_entries
+      @references ||= []
+    end
+
+    def entries_from_plugins
+      Foreman::Plugin.all.flat_map(&:smart_proxy_references).compact.reduce(:merge)
```

Comment:
This can cause problems, since it overwrites the column list:
``` ruby
[{:self => [:id] },{ :self => [:name]}].reduce(:merge)
# => {:self=>[:name]}
```